packages/dds/tree/src/simple-tree/api/identifierIndex.ts

```diff
@@ -0,0 +1,270 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import type {
+	AnchorNode,
+	FieldKey,
+	ITreeSubscriptionCursor,
+	TreeNodeSchemaIdentifier,
+} from "../../core/index.js";
+// todo fix this
+// eslint-disable-next-line import/no-internal-modules
+import { makeTree } from "../../feature-libraries/flex-tree/lazyNode.js";
+import {
+	type Context,
+	AnchorTreeIndex,
+	isTreeValue,
+	type TreeIndexNodes,
+	hasElement,
+	type TreeIndex,
+	type TreeIndexKey,
+	type KeyFinder,
+} from "../../feature-libraries/index.js";
+import { brand, fail } from "../../util/index.js";
+import type { ImplicitFieldSchema, NodeFromSchema } from "../schemaTypes.js";
+import {
+	getOrCreateNodeFromInnerNode,
+	type TreeNode,
+	type TreeNodeSchema,
+} from "../core/index.js";
+import { ObjectNodeSchema } from "../objectNodeTypes.js";
+import { treeNodeApi } from "./treeNodeApi.js";
+// todo
+// eslint-disable-next-line import/no-internal-modules
+import { proxySlot } from "../core/treeNodeKernel.js";
+import type { TreeView } from "./tree.js";
+import { walkFieldSchema } from "../walkFieldSchema.js";
+// todo
+// eslint-disable-next-line import/no-internal-modules
+import type { SchematizingSimpleTreeView } from "../../shared-tree/schematizingTreeView.js";
+
+/**
+ * A {@link TreeIndex} that returns tree nodes given their associated keys.
+ *
+ * @alpha
+ */
+export type SimpleTreeIndex<TKey extends TreeIndexKey, TValue> = TreeIndex<TKey, TValue>;
+
+/**
+ * Creates a {@link SimpleTreeIndex} with a specified indexer.
+ *
+ * @param view - the view for the tree being indexed
+ * @param indexer - a function that takes in a {@link TreeNodeSchema} and returns the field name that all nodes of the given schema should be keyed on
+ * @param getValue - given at least one {@link TreeNode}, returns an associated value
+ * @param isKeyValid - function for verifying the validity of the key retrieved based on the information given by the indexer
+ *
+ * @alpha
+ */
+export function createSimpleTreeIndex<
```

Comment:
Nondeterminism isn't an issue here because the function is only run once for a particular input schema. I added to the doc comment that this function should be pure and functional as well.

This API isn't super flexible because it still only allows keys that are one level underneath the particular nodes to be indexed. We did choose to expose it this way because it allows for some cleaner definitions. For example, if you have a particular key you want to index on for many different schemas, instead of defining a mapping of [[Tree, "foo"], [Node, "foo"], [Leaf, "foo"], ..., [Etc, "foo"]], you could define just a function (schema) => if schema has field foo return "foo". I added the map version as an overload so that whichever makes sense can be used.